src/core/processors/poolprocessor.cpp

```diff
@@ -0,0 +1,142 @@
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <inviwo/core/processors/poolprocessor.h>
+#include <inviwo/core/network/processornetwork.h>
+#include <inviwo/core/common/inviwoapplication.h>
+
+namespace inviwo {
+
+void pool::Progress::operator()(float progress) const noexcept {
+    state_.setProgress(id_, progress);
+}
+
+void pool::Progress::operator()(double progress) const noexcept {
+    state_.setProgress(id_, static_cast<float>(progress));
+}
+
+void pool::detail::State::setProgress(size_t id, float newProgress) {
+    IVW_ASSERT(id < progress.size(), "Invalid job id");
+
+    progress[id] = newProgress;
+
+    if (!progressUpdate.valid() || util::is_future_ready(progressUpdate)) {
+        const auto total =
+            std::accumulate(progress.begin(), progress.end(), 0.0f) / progress.size();
+        progressUpdate = dispatchFront([this, p = processor, total]() {
+            if (auto wrapper = p.lock()) {
+                wrapper->processor.progress(this, total);
+            }
+        });
+    }
+}
+
+PoolProcessor::PoolProcessor(pool::Options options, const std::string& identifier,
+                             const std::string& displayName)
+    : Processor(identifier, displayName)
+    , options_{options}
+    , wrapper_{std::make_shared<pool::detail::Wrapper>(*this)}
+    , delay_{std::chrono::milliseconds(500),
+             [wrapper = std::weak_ptr<pool::detail::Wrapper>(wrapper_)]() {
+                 if (auto w = wrapper.lock()) {
+                     auto& p = w->processor;
+
+                     if (p.queuedDispatch() && !p.states_.empty()) return;
+
+                     if (!p.queue_.empty()) {
+                         p.submit(p.queue_.front());
+                         p.queue_.clear();
+                     }
+                 }
+             }} {}
+
+PoolProcessor::~PoolProcessor() { stopJobs(); }
+
+void PoolProcessor::stopJobs() {
+    for (auto& state : states_) {
+        state->stop = true;
+    }
+}
+
+bool PoolProcessor::hasJobs() { return !states_.empty(); }
+
+void PoolProcessor::submit(Submission& job) {
+    job.setupProgress();
+    states_.push_back(job.state);
+    for (auto& task : job.tasks) {
+        getNetwork()->getApplication()->getThreadPool().enqueueRaw(std::move(task));
+    }
+}
+
+void PoolProcessor::invalidate(InvalidationLevel invalidationLevel, Property* source) {
+    if (delayInvalidation()) {
+        notifyObserversInvalidationBegin(this);
+        PropertyOwner::invalidate(invalidationLevel, source);
+        notifyObserversInvalidationEnd(this);
+    } else {
+        Processor::invalidate(invalidationLevel, source);
+    }
+}
+
+void PoolProcessor::handleError() {
```

Comment:
Is there any way to get information about the error?

Reply:
just rethrow it

Reply:
Rethrow what? There is no info passed into the function. Seems a bit magic to me :)

Reply:
You can rethrow any caught exception

Reply:
We use that in quite a few places to delegate error handling